src/cockpit/beiboot.py

```diff
@@ -129,10 +129,25 @@ class AuthorizeResponder(ferny.AskpassHandler):
     commands = ('ferny.askpass', 'cockpit.report-exists')
     router: Router
 
-    def __init__(self, router: Router):
+    def __init__(self, router: Router, basic_password: Optional[str]):
         self.router = router
+        self.basic_password = basic_password
+        self.have_basic_password = bool(basic_password)
 
     async def do_askpass(self, messages: str, prompt: str, hint: str) -> Optional[str]:
+        logger.debug("AuthorizeResponder: prompt %r, messages %r, hint %r", prompt, messages, hint)
+
+        if self.have_basic_password and 'password:' in prompt.lower():
+            # the UI currently expects us to fail on a wrong password, so only send it once
+            if self.basic_password is not None:
```

Comment:
This confused the heck out of me until I realized that you were setting `self.basic_password = None` below.

As far as I can tell, the code is correct as it's written, but the `have_basic_password` should definitely be renamed to something like `had_` or `got_` to remove the suggestion that it applies to the current state.

I might also just drop that and do something like

```python
if self.basic_password and 'password:' in prompt.lower():
    if self.password_attempted:
        raise ...

    self.password_attempted = True
    return self.basic_password
```

It has the disadvantage of not "freeing" the password, but you're not really scrubbing it anyway...

Reply:
Wrt. the renaming: it may be the current state, just that have_basic_password stays (to remember in which "mode" we are), and basic_password gets consumed, as it doesn't make sense to try more than once. `got_` is certainly fine, but if you think `password_attempted` is cleaner to read, that's fine for me.

Scrubbing in Python is indeed difficult, as `str` is immutable AFAIK. I.e. the only thing I can hope for is that the memory gets GCed and eventually overwritten. This may be possible with `StringIO` or similar, but I figure all of this is kind of hopeless in Python, and we rather rely on ptrace protection and such..